opt/optbuilder: support ORDER BY in the optbuilder

rytaft opened the pull request:
This commit builds on #23123 to support all types of
`ORDER BY` expressions in the optbuilder, not just `ORDER BY`
expressions containing variables found in the `SELECT` list.

For example, it is possible to order by column aliases,
`GROUP BY`/`FROM` columns that are not in the `SELECT` list,
scalar expressions using these `GROUP BY`/`FROM` columns,
ordinal indexes into the `SELECT` list, etc.

Some examples of supported queries are:
```
  SELECT a FROM t ORDER BY b, c
  SELECT a, b, a+b AS ab FROM t ORDER BY ab DESC
  SELECT * FROM t ORDER BY t.*
  SELECT a, b, c FROM t ORDER BY 2 DESC
```

The test cases have been adapted from the order_by tests
in logictests.

This commit does NOT include support for `ORDER BY INDEX idx@tbl`
 or `ORDER BY PRIMARY KEY`. Some optimizations are also left for a
future PR. For example,
```
  SELECT a+b FROM t ORDER BY a+b
```
causes two projections to be built for `a+b`. A future PR
should perform de-duplication.

This code is based on code originally written as part of the
opttoy project.

Release note: None

rytaft committed:
opt/optbuilder: support ORDER BY in the optbuilder

This commit builds on #23123 to support all types of
ORDER BY expressions in the optbuilder, not just ORDER BY
expressions containing variables found in the SELECT list.

For example, it is possible to order by column aliases,
GROUP BY/FROM columns that are not in the SELECT list,
scalar expressions using these GROUP BY/FROM columns,
ordinal indexes into the SELECT list, etc.

Some examples of supported queries are:
  SELECT a FROM t ORDER BY b, c
  SELECT a, b, a+b AS ab FROM t ORDER BY ab DESC
  SELECT * FROM t ORDER BY t.*
  SELECT a, b, c FROM t ORDER BY 2 DESC

The test cases have been adapted from the order_by tests
in logictests.

This commit does NOT include support for ORDER BY INDEX
idx@tbl or ORDER BY PRIMARY KEY. Some optimizations are
also left for a future PR. For example,
  SELECT a+b FROM t ORDER BY a+b
causes two projections to be built for a+b. A future PR
should perform de-duplication.

This code is based on code originally written as part of the
opttoy project.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23554)
<!-- Reviewable:end -->

rytaft commented:
Thanks for the reviews! Please take another look.

---

Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/column_props.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71BqYDCuULe2bnxEun:-L75KGrnINA1hqwdPFdh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/column_props.go#L46)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

did you want to set c.exprStr here?
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/orderby.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71mtb2XfHWSDmz8p7T:-L75KL8U5ibjT68p-os9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/orderby.go#L59)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: violation of 100 char line limit causes side-by-side wrapping.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/orderby.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71tKUaLxSqiFpCQZJr:-L75KMtHhKMPQuLGi_xM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/orderby.go#L151)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This is a fairly generic name for something that documented as ORDER BY specific. I'd name it more specifically to prevent confusion.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/project.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71saVzwhlvuOV4Haq0:-L75KO0TamM0OhVqkzoK:b-xumyc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/project.go#L205)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I've got concerns about sprinkling these assignments in multiple places around the builder:

1. It cause fragility, because it's not clear when this should happen, and what the string should be.
2. It seems to be overlapping with "label". Even if it isn't really overlapping, it will cause confusion as-is, because developers will think it is.
3. It's an expensive operation. I think we should be storing the expression and then lazily get its string only when necessary.

Do some careful thinking about what all the different fields are, and how to prevent other developers from coming in later on and being confused, or forgetting to set `exprStr` when they should have set it, or setting it to the wrong thing.
</blockquote></details>

All good points. I've added fields in `columnProps` for the original typed expression, as well as the original column name. `exprStr` is lazily evaluated by `getExprStr`.

---

*[pkg/sql/opt/optbuilder/testdata/orderby, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71jso7fd4bOXrGJVfw:-L75KnJ-vwhMWI6cWKy1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/testdata/orderby#L91)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this should be max not min (because of the DESC)
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/testdata/orderby, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L71jh0czY4233kDWbR3:-L75KoeaSVZxvmhz7g43:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b8403d5cedeedaa18683a7683c3a34d3868d31/pkg/sql/opt/optbuilder/testdata/orderby#L96)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this error message is not ideal (pg error is `for SELECT DISTINCT, ORDER BY expressions must appear in select list`), maybe add a TODO
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-:-L75Ks3o6FNqoDykatkv:bm8vf4d)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 16 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/column_props.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75Q_WX9q48RWLbqRHt:-L75Q_WX9q48RWLbqRHu:bx4k9vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/column_props.go#L30)):*
> ```Go
> // replaced with a columnProps.
> type columnProps struct {
> 	// origName is the original name of this column.
> ```

NIT: a bit more detail would be nice, like `is the original name of this column, either in its origin table, or when it was first synthesized.`

---

*[pkg/sql/opt/optbuilder/join.go, line 277 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75R_Oni5HuWspHimnD:-L75R_Oni5HuWspHimnE:b15znvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/join.go#L277)):*
> ```Go
> 			// Construct a new merged column to represent IFNULL(left, right).
> 			texpr := tree.NewTypedCoalesceExpr(tree.TypedExprs{leftCol, rightCol}, leftCol.typ)
> 			col := b.synthesizeColumn(outScope, string(leftCol.name), leftCol.typ, texpr)
> ```

Should `texpr.ResolvedType()` be passed to `synthesizeColumn`? Like what if `leftCol.typ` is `types.Null`?

---

*[pkg/sql/opt/optbuilder/orderby.go, line 66 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75Rz9XFzuwwAwNqnXI:-L75Rz9XFzuwwAwNqnXJ:b-rptxvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/orderby.go#L66)):*
> ```Go
> }
> 
> func (b *Builder) buildOrder(
> ```

NIT: Maybe `buildOrderings`?

---

*[pkg/sql/opt/optbuilder/select.go, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75SJYT7REuOiNxm3zX:-L75SJYT7REuOiNxm3zY:b-7dxsxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/select.go#L179)):*
> ```Go
> 		projections := make([]opt.GroupID, 0, len(outScope.cols))
> 		for i := range outScope.cols {
> 			projections = append(projections, b.buildScalarProjection(&outScope.cols[i], "", outScope, projectionsScope))
> ```

NIT: violation of 100-char line limit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-:-L75SgO6cfGSM7fVUTON:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks! Updated.

---

Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/optbuilder/column_props.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75Q_WX9q48RWLbqRHt:-L76P6mK3mYfeNlX4QJf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/column_props.go#L30)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: a bit more detail would be nice, like `is the original name of this column, either in its origin table, or when it was first synthesized.`
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/join.go, line 277 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75R_Oni5HuWspHimnD:-L76Hg3gPyi-k9s4Ukqk:brqbmt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/join.go#L277)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should `texpr.ResolvedType()` be passed to `synthesizeColumn`? Like what if `leftCol.typ` is `types.Null`?
</blockquote></details>

Good point!  Added some logic to check the type, since in this case `texpr.ResolvedType()` will just return the type passed to `NewTypedCoalesceExpr`.  Also added a new test case for this issue.

(Note that this doesn't work in CockroachDB today - filed issue #23609.)

---

*[pkg/sql/opt/optbuilder/orderby.go, line 66 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75Rz9XFzuwwAwNqnXI:-L76Q1lnlQhX08Llooov:bd990pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/orderby.go#L66)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Maybe `buildOrderings`?
</blockquote></details>

Done. (I made it singular since it's just handling one of the Order By columns)

---

*[pkg/sql/opt/optbuilder/select.go, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-L75SJYT7REuOiNxm3zX:-L76Q9G8QM0a9QC8MTZM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a2add947998cb473c081c2254c8822ef505ff96/pkg/sql/opt/optbuilder/select.go#L179)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: violation of 100-char line limit
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23554#-:-L76QDhNdQaonojkQY0y:b-wb3jnk)*
<!-- Sent from Reviewable.io -->
